lib/promscrape/client.go

```diff
@@ -364,16 +365,23 @@ func doRequestWithPossibleRetry(hc *fasthttp.HostClient, req *fasthttp.Request,
 		} else if err != fasthttp.ErrConnectionClosed && !strings.Contains(err.Error(), "broken pipe") {
 			return err
 		}
-		// Retry request after exponentially increased sleep.
-		maxSleepTime := time.Until(deadline)
-		if sleepTime > maxSleepTime {
-			return fmt.Errorf("the server closes all the connection attempts: %w", err)
-		}
-		sleepTime += sleepTime
-		if sleepTime > maxSleepTime {
-			sleepTime = maxSleepTime
+
+		// Special handling for cases when server closes the connection.
+		// This might be caused by connection keep-alive timeout being exceeded.
+		// In such cases additional delay will not give any benefit, so retry immediately.
+		if !firstRetry {
```

Comment:
You sure about `!`?

Reply:
You set firstRetry to true and it never switches to false :)

Reply:
It does on the line 373 :)

Reply:
It would never go into that line - that's why I asked. You are setting it to true, so it never goes into `if` and never can become `false` to go into that if. Now the logic is simplified, so it doesn't matter anymore :)